include/clap/preset-discovery.h

```diff
@@ -0,0 +1,244 @@
+
+   // gets the location info at index
+   bool(CLAP_ABI *get_location)(const struct clap_preset_provider *provider,
+                                uint32_t                           index,
+                                clap_preset_location_t            *out_info);
+
+   // returns the number of different supported file extensions
+   uint32_t(CLAP_ABI *file_extensions_count)(const struct clap_preset_provider *provider);
+
+   // stores the file extensions at index into out_extension.
+   // the `.' isn't included in the string.
+   // extension_capacity indicates the number of bytes available to store the extension
+   bool(CLAP_ABI *get_file_extension)(const struct clap_preset_provider *provider,
+                                      uint32_t                           index,
+                                      char                              *out_extension,
+                                      uint32_t                           extension_capacity);
+
+   // gets information about a given collection.
+   bool(CLAP_ABI *get_collection_info)(const struct clap_preset_provider *provider,
+                                       const char                        *collection_id,
+                                       clap_preset_collection_info_t     *out_info);
+
+   // reads metadata from the given file and passes them to the metadata receiver
+   bool(CLAP_ABI *read_file_metadata)(const struct clap_preset_provider     *provider,
+                                      const char                            *file_path,
+                                      const clap_preset_metadata_receiver_t *metadata_receiver);
+} clap_preset_provider_t;
+
+typedef struct clap_preset_indexer {
+   clap_version_t clap_version; // initialized to CLAP_VERSION
+   const char    *name;
+   const char    *vendor;
+   const char    *version;
+} clap_preset_indexer_t;
+
+static const CLAP_CONSTEXPR char CLAP_PRESET_DISCOVERY_FACTORY_ID[] =
+   "clap.preset-discovery-factory";
+
+// Every methods in this factory must be thread-safe.
+// It is encourraged to perform preset indexing in background threads, maybe even in background
+// process.
+typedef struct clap_preset_discovery_factory {
+   // Get the number of preset providers available.
+   // [thread-safe]
+   uint32_t(CLAP_ABI *count)(const struct clap_preset_discovery_factory *factory);
+
+   // Retrieves a preset provider descriptor by its index.
+   // Returns null in case of error.
+   // The descriptor must not be freed.
+   // [thread-safe]
+   const clap_preset_provider_descriptor_t *(CLAP_ABI *get_descriptor)(
+      const struct clap_preset_discovery_factory *factory, uint32_t index);
+
+   // Create a preset provider by its id.
+   // The returned pointer must be freed by calling preset_provider->destroy(preset_provider);
+   // The preset provider is not allowed to use the indexer callbacks in the create method.
+   // Returns null in case of error.
```

Comment:
What are the 'indexer callbacks'? clap_preset_indexer_t doesn't have any functions.

Reply:
for now it doesn't have, maybe it could have some I'm not sure yet.